README.md
=====================
vdr-plugin-squeezebox
=====================

Copyright:            Jörg Wendel / horchi (vdr@jwendel.de)
Project's homepage:   https://github.com/horchi/vdr-plugin-squeezebox
Latest version at:    git clone https://github.com/horchi/vdr-plugin-squeezebox

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
See the file COPYING for more information.

The basic OSD drwaing function taken from skin-nopacity. Thanks to Louis Braun!

Description:
-------------

Plugin which act as a soft squeezebox


Requirements:
-------------

- libmagick++-dev
- running and reachable LMC in your network (see below)
- squeezelite (see below)
- a output device wich supports true color OSD

Installation:
-------------

1. Plugin
  
  #> git clone https://github.com/horchi/vdr-plugin-squeezebox
  #> cd vdr-plugin-squeezebox
  #> make
  #> make install

  may be you have to fix the access rights of <plugin-conf-path>/squeezebox 


2. softhddevice Patch

  The squeezebox plugin need a deblocked sound device, a blacked video 
  but avalible OSD output by the output device plugin. Exactly as I understand 
  the play mode 'ePlayMode = pmAudioOnlyBlack' (audio only from player, no video (black screen)).

  I use the softhddevice plugin in version 0.6.1rc1 as output plugin, to enable the pmAudioOnlyBlack support in the softhddevice
  the patch softhddevice-pmAudioOnly.patch (deliverd with the squeezebox plugin) is needed! 
  Therefore you habe to apply this patch and recompile softhddevice. Johns like to adopt the patch in one of the next versions.
  
  I don't know the behavior of pmAudioOnlyBlack relating to other output plugins like xine ...


3. LMS - Logitech Media Server

  Install the LMS on your server

    Logitech Version: http://www.mysqueezebox.com/download
    or the fork:      http://downloads.slimdevices.com/nightly/ (ich verwende aktuell die 7.8.0)

  Configure it via the WEB interface

4. squeezelite - streaming client for the LMS

  squeezelite is the the streming client used by the plugin
  Requirements: libasound2-dev libflac-dev libmad0-dev libvorbis-dev libfaad-dev libmpg123-dev

  #> git clone https://code.google.com/p/squeezelite
  #> cd squeezelite
  #> make
  #> cp squeezelite /usr/local/bin/squeezelite
  #> chmod 755 /usr/local/bin/squeezelite

Configuration:
--------------

  Put this in the setup.conf (until the plugins setup dialog is implemented):

   squeezebox.audioDevice = hw:CARD=NVidia,DEV=9
   squeezebox.lmcHost = localhost
   squeezebox.lmcHttpPort = 9000
   squeezebox.lmcPort = 9090
   squeezebox.logLevel = 1
   squeezebox.playerName = VDR-squeeze
   squeezebox.squeezeCmd = /usr/local/bin/squeezelite

  And adjust the parameters to fit your system:

  - squeezebox.audioDevice  
     The possible setting of audioDevice are shown by calling:
     #> squeezelite -l
     normally similar to the softhddevice configuration (squeezelite option)
   
  - squeezebox.logLevel
     the loglevel (0-4), 0 less, 1 normal, 2.. debug (default 1)

  - squeezebox.playerName 
     name of the squeezebox as displayed in the WEBIF (squeezelite option, default 'VDR-squeeze')

  - squeezebox.playerMac
     mac of the squeezebox, to run more than one squueze clinet in the same host (default the real mac)

  - squeezebox.squeezeCmd:
     path of squeezelite (default /usr/local/bin/squeezelite)

  - squeezebox.lmcHost
    host or ip of the LMS (default localhost)

  - squeezebox.lmcHttpPort
    http port of the LMS (default 9000)

  - squeezebox.lmcPort
    port of LMSs CLI interface (default 9090)

  - squeezebox.shadeTime
    inactivity time to shade the OSD for off (default 0)

  - squeezebox.shadeLevel
    shade level 0-100 in %, 100 is black (default 40)
   
    
Handling:
---------

Should intuitive ;)

- key 0 changes the color button setting


HISTORY

VDR Plugin 'squeezebox' Revision History
----------------------------------------

2014-02-22: Version 0.0.13
  - added: alsa options to configuration
  - bugfix: storage of log level (by alexander grothe)

2014-02-18: Version 0.0.12
  - added: delete of player instance on plugin exit

2014-02-17: Version 0.0.11
  - added  Setup menu
  - added  auto scroll for lyrics 
  - added  shadeLevel (0-100) to config

2014-02-02: Version 0.0.10
  - change  fixed spacing below cover box
  - change  added lyrics 
  - bugfix  waiting on player startup

2014-01-30: Version 0.0.9
  - added   station play time for webradio instead of track time/duration
  - added   'favorites' to menu
  - added   'new music' to menu
  - bugfix  fixed missing PNG error
  - added   support of VDRs 'MenuScrollWrap' setting
  - added   shade on inacivity (use README for details)
  - added   start replay on plugin start
  - added   dimmed cover as menu background

2014-01-28: Version 0.0.8
  - added   key repead
  - added   support of long artist, genre, ... lists
  - added   cover art for web radio
  - change  cover load without fs usage (speed)
  - added   track info for webradio (station name, ...)
  - change  improved scroll speed of playlist
  - added   menue for web radio
  - added   TODO list
  - added   mac to configuration (setup.conf)
  - added   artwork cache to speedup redraw

2014-01-25: Version 0.0.7
  - added   own menu instead of skins-menu

2014-01-24: Version 0.0.6
  - added   menu structure like LMCs WEBIF
  - change  PNGs stored in resdir (instead of cfgdir)
  - added   DEST target to make install

2014-01-20: Version 0.0.5
  - added   osd redraw on vdr volume dispay

2014-01-20: Version 0.0.4
  - changed  color button radius
  - change  view of progress bar even replay is paused

2014-01-20: Version 0.0.3
  - First release

2014-01-12: Version 0.0.2
  - First working beta.

2014-01-02: Version 0.0.1
  - Initial revision.


config.c
/*
 * config.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 */

#include <string.h>

#include "lib/common.h"
#include "config.h"

//***************************************************************************
// Config
//***************************************************************************

cSqueezeConfig::cSqueezeConfig()
{
   logLevel = 1;
   lmcHost = strdup("localhost");
   lmcPort = 9090;
   lmcHttpPort = 9000;

   squeezeCmd = strdup("/usr/local/bin/squeezelite");
   playerName = strdup("VDR-squeeze");
   audioDevice = strdup("");
   alsaOptions = strdup("");

   shadeTime = 0;
   shadeLevel = 40;  // in %

   mac = getMac();
}

cSqueezeConfig::~cSqueezeConfig()
{
   free(lmcHost);
   free(mac);
   free(squeezeCmd);
   free(playerName);
   free(audioDevice);
   free(alsaOptions);
}

//***************************************************************************
// Setup Menu
//***************************************************************************

cMenuSqueezeSetup::cMenuSqueezeSetup() 
{
   strcpy(lmcHost, cfg.lmcHost);
   strcpy(squeezeCmd, cfg.squeezeCmd);
   strcpy(playerName, cfg.playerName);
   strcpy(audioDevice, cfg.audioDevice);
   strcpy(alsaOptions, cfg.alsaOptions);
   strcpy(mac, cfg.mac);
   
   Setup();
}

void cMenuSqueezeSetup::Setup()
{  
   Clear();

   Add(new cMenuEditStrItem(tr("LMS Host"), lmcHost, sizeof(lmcHost), tr(FileNameChars)));
   Add(new cMenuEditIntItem(tr("LMS Port"), &cfg.lmcPort, 1, 99999));
   Add(new cMenuEditIntItem(tr("LMS/Http Port"), &cfg.lmcHttpPort, 1, 99999));

   Add(new cMenuEditStrItem(tr("Player Name"), playerName, sizeof(playerName), tr(FileNameChars)));
   Add(new cMenuEditStrItem(tr("Player MAC"), mac, sizeof(mac), tr(FileNameChars)));
   Add(new cMenuEditStrItem(tr("Squeezelite Path"), squeezeCmd, sizeof(squeezeCmd), tr(FileNameChars)));
   Add(new cMenuEditStrItem(tr("Audio Device"), audioDevice, sizeof(audioDevice), tr(FileNameChars)));
   Add(new cMenuEditStrItem(tr("Alsa Options"), alsaOptions, sizeof(alsaOptions), tr(FileNameChars)));

   Add(new cMenuEditIntItem(tr("Shade Time [s]"), &cfg.shadeTime, 0, 3600));
   Add(new cMenuEditIntItem(tr("Shade Level [%]"), &cfg.shadeLevel, 0, 100));

   Add(new cMenuEditIntItem(tr("Log level"), &cfg.logLevel, 0, 4));

//   SetCurrent(Get(current));
   Display();
}

eOSState cMenuSqueezeSetup::ProcessKey(eKeys Key) 
{
   eOSState state = cMenuSetupPage::ProcessKey(Key);
   
   switch (state) 
   {
      case osContinue:
      {
         if (NORMALKEY(Key) == kUp || NORMALKEY(Key) == kDown) 
         {
            cOsdItem* item = Get(Current());

            if (item)
               item->ProcessKey(kNone);
         }

         break;
      }
         
      default: break;
   }

   return state;
}

void cMenuSqueezeSetup::Store(void)
{
   free(cfg.lmcHost);     cfg.lmcHost = strdup(lmcHost); 
   free(cfg.squeezeCmd);  cfg.squeezeCmd = strdup(squeezeCmd);  
   free(cfg.playerName);  cfg.playerName = strdup(playerName);  
   free(cfg.mac);         cfg.mac = strdup(mac);   
   free(cfg.audioDevice); cfg.audioDevice = strdup(audioDevice); 
   free(cfg.alsaOptions); cfg.alsaOptions = strdup(alsaOptions); 

   SetupStore("logLevel", cfg.logLevel);
   SetupStore("lmcHost", cfg.lmcHost);
   SetupStore("lmcPort", cfg.lmcPort);
   SetupStore("lmcHttpPort", cfg.lmcHttpPort);
   SetupStore("shadeTime", cfg.shadeTime);
   SetupStore("shadeLevel", cfg.shadeLevel);
   SetupStore("squeezeCmd", cfg.squeezeCmd);
   SetupStore("playerName", cfg.playerName);
   SetupStore("playerMac", cfg.mac);
   SetupStore("audioDevice", cfg.audioDevice);
   SetupStore("alsaOptions", cfg.alsaOptions);
}
